doc/designs/fips_indicator.md

```diff
@@ -0,0 +1,222 @@
+- Documentation is required to demonstrate how to use indicators for each approved cryptographic algorithm.
+- Testing is required to execute all services and verify that the indicator provides an unambiguous indication of whether the service utilizes an approved cryptographic algorithm, security function or process in an approved manner or not.
+- The Security Policy may require updates related to indicators. AWS/google have added a table in their security policy called ‘Non-Approved Algorithms not allowed in the approved mode of operation’. An example is RSA with a keysize of < 2048 bits (which has been enforced by [4]).
+
+Since any new FIPS restrictions added could possibly break existing applications
+the following additional OpenSSL requirements are also needed:
+
+- The FIPS restrictions should be able to be disabled using Configuration file options (This results in unapproved mode and requires an indicator).
+- A mechanism for logging the details of any unapproved mode operations that have been triggered (e.g. DSA Signing)
+- The FIPS restrictions should be able to be enabled/disabled per algorithm context.
+- If the per algorithm context value is not set, then the  Configuration file option is used.
+
+Solution
+--------
+
+In OpenSSL most of the existing code in the FIPS provider is using
+implicit indicators i.e. An error occurs if existing FIPS rules are violated.
+
+The following rules will apply to any code that currently is not FIPS approved,
+but needs to be.
+
+- The fipsinstall application will have a configurable item added for each algorithm that requires a change.
+These options will be passed to the FIPS provider in a manner similar to existing code.
+
+- A user defined callback similar to OSSL_SELF_TEST will be added. This callback
+will be triggered whenever an approved mode test fails.
+It may be set up by the user using
+
+```c
+OSSL_INDICATOR_set_callback(OSSL_LIB_CTX *libctx, OSSL_CALLBACK *cb, void *cbarg)
+```
+
+- The callback API will be
+
+```c
+int OSSL_INDICATOR_callback(OSSL_LIB_CTX *libctx, const char *algtype,
+                            const char *algdesc)
+```
+
+This can be used during application testing to log that an indicator was
+triggered. The user callback may return 0 if the application wants an error
+to occur based on the indicator type and description.
+
+- To control an algorithm context's checks via code requires a setter (e.g OSSL_ALG_PARAM_STRICT_CHECKS),
+
+```c
+    p = OSSL_PARAM_locate_const(params, OSSL_ALG_PARAM_STRICT_CHECKS);
+    if (p != NULL
+        && !OSSL_PARAM_get_int(p, &ctx->strict_checks))
+        return 0;
+```
+
+The setter is initially -1 (unknown) and can be set to 0 or 1 via a set_ctx call.
```

Comment:
I'm not sure a tri-state here is sensible.  I'm also not sure that configuration should be overridden by this.

Reply:
The -1 state allows the code to determine if the value has been set or NOT. See the dsa PR for why this was used.